### DISCUSSION TITLE
How to insert a row from a custom endpoint?

### DISCUSSION THREAD

**Discussion by glowkeeper** [directus/directus#4056] [Archive]
How do you add a row to a collection from a custom endpoint? 

The docs give a query as an example:
```
module.exports = function registerEndpoint(router, { services, exceptions }) {
	const { ItemsService } = services;
	const { ServiceUnavailableException } = exceptions;

	router.get('/', (req, res) => {
		const recipeService = new ItemsService('recipes', { schema: req.schema });
		recipeService
			.readByQuery({ sort: 'name', fields: ['*'] })
			.then((results) => res.json(results))
			.catch((error) => {
				throw new ServiceUnavailableException(error.message);
			});
	});
};
```
...and tell me, `services — All API internal services`; hence, I'm guessing (hoping!) that one of the services is an _insert_. Fingers crossed that I'll be able to do something like:
```
router.get('/', (req, res) => {
		const myCoolService = new ItemsService('foo', { schema: req.schema });
		myCoolService
			.insert({ 'bar': 'alice', 'key': 'bob' })
			.then((results) => res.json(results))
			.catch((error) => {
				throw new ServiceUnavailableException(error.message);
			});
	});
```
Am I heading in the right direction?

**Comment by glowkeeper** [comment]
I'm guessing this is a _knex_ something or other...I'll get there ;)

**Comment by rijkvanzanten** [comment]
It's `create`, but other than that you're spot on 🙂 

See:

https://github.com/directus/directus/blob/main/api/src/services/items.ts#L48

**Reply by glowkeeper** [reply]
Oooo - I found that file, but I was looking for `insert`...

**Reply by glowkeeper** [reply]
Eek! Help! I nearly got it working, but not quite...

I have a collection, _wallet_, with an auto-generated primary key (`id`), and two other fields: `userid` (which is a many-to-one field that points back to `directus_users`), and `publickey`, which is a text field. The _public_ user has read/write permissions to _wallet_.

I tried this:
```
        router.post('/key', (req, res) => {

               const walletService = new ItemsService('wallet', { schema: req.schema });
		walletService
			.create( {'userid': req.body.id, 'publickey': req.body.key})
			.then((results) => res.json(results))
			.catch((error) => {
				throw new ServiceUnavailableException(error.message);
			});

                return (
                        res.send("OK")
                )
	});
```
And I called it with this:
```
curl --header "Content-Type: application/json"   --request POST   --data '{"id":"813a4267-a9e7-4949-8044-18ded8782c95", "key":"sdfgsdfg"}' http://localhost:8055/custom/minima/key
```
It _partially_ worked - it created a new row, but I got _UnhandledPromiseRejectionWarning: Error: Cannot set headers after they are sent to the client_, and it didn't set anything for the `userid` field (it successfully added the `publickey`, though). And even though I'm logged in as Admin to the admin app', when I try and view the new row, it tells me I'm _forbidden_ (screengrab attached). 

<img width="1440" alt="Screenshot 2021-02-15 at 16 54 22" src="https://user-images.githubusercontent.com/9264929/107974539-a816ef00-6fae-11eb-9d39-14a31b2e9247.png">

Apologies - I must be missing something obvious...

**Reply by glowkeeper** [reply]
Aha! if I make `userid` just a standard text field (and not a Many-to-one relationship thingy that points back to `directus_users`), the row gets created successfully. I still get _UnhandledPromiseRejectionWarning: Error: Cannot set headers after they are sent to the client_, but I'll work that out...

**Reply by glowkeeper** [reply]
(got rid of the _ UnhandledPromiseRejectionWarning_ by removing the `return`. Silly me)

**Reply by rijkvanzanten** [reply]
> UnhandledPromiseRejectionWarning: Error: Cannot set headers after they are sent to the client

This is because you're trying to send the results with `res.json` _after_ you've already sent the string "OK".

That forbidden error is most likely the relational interface trying to read an item that doesn't exist. Are you sure your provided `userid` is correct? Are you not accidentally saving the `publickey` in `userid` or something like that?

**Reply by glowkeeper** [reply]
Gosh - I'm embarrassed - it was working all along! I copied a `userid` from another instance of _directus_ that I have running on a server - oops! 

Sometimes I hate hanging dirty washing out in public, but all's well that ends well - thanks for your help @rijkvanzanten !

**Reply by glowkeeper** [reply]
Actually, related to this - what's the easiest way to check if a supplied `userid` is a valid `directus_users` `id`?

**Reply by rijkvanzanten** [reply]
The most reliable way is trying to read it from the UsersService.readByKey. If that returns, it exists

**Reply by glowkeeper** [reply]
I'm trying this, and failing miserably:
```
const userService = new UsersService({ schema: req.schema })
      userService
        .readByKey(`${userid}`)
```
:(

**Reply by rijkvanzanten** [reply]
Are you defining UsersService the same as ItemsService from before?

```js
const { ItemsService } = services;
```

becomes

```js
const { ItemsService, UsersService } = services;
```

(btw, ``(`${userid}`)`` would be the same as `(userid)` 🙂 )

**Reply by glowkeeper** [reply]
> const { ItemsService, UsersService } = services;

yes

**Reply by glowkeeper** [reply]
```js
module.exports = function registerEndpoint(router, { services, exceptions }) {

  const { ItemsService, UsersService } = services;
  const { ServiceUnavailableException } = exceptions;

  router.post('/key', (req, res) => {

    const userid = req.body.userid
    const publickey = req.body.publickey

    if ( userid && publickey ) {

      const userService = new UsersService({ schema: req.schema })
      userService
        .readByKey(`${userid}`, {fields: ['*']})
        .then((results) => {
```
That's the gist of it :)

**Reply by rijkvanzanten** [reply]
What you have here looks good to me

**Reply by glowkeeper** [reply]
My, "doesn't work!" wasn't especially helpful - apologies :(

So the error I'm actually seeing is: **You don't have permission to access this.** 

Just for testing, I've made `directus_users` readable and writeable by _everyone_ (Public, User et al.), so I'm surprised I don't have permissions. 

I'm using _cURL_ for testing purposes, like so:
```
curl --header "Content-Type: application/json" --request POST --data '{"userid":"foo","publickey":"bar"}' http://localhost:8055/custom/minima/key
```
Is it something to do with _cURL_? Or do I have to set more permissions somewhere?

**Reply by glowkeeper** [reply]
Actually - it can't be _cURL_, because this seems to run successfully, via _cURL_:
```
router.post('/key', (req, res, next) => {

    const userid = req.body.userid;
    const publickey = req.body.publickey;

    if ( userid && publickey ) {

                const walletService = new ItemsService('wallet', { schema: req.schema });
      		walletService
      			.create({'userid': userid, 'publickey': publickey})
      			.then((results) => {
```
Whereas this, via cURL, gets told: **You don't have permission to access this**:
```
router.get('/key', (req, res, next) => {

    const userid = req.body.userid

    if ( userid ) {

      const walletService = new ItemsService('wallet', { schema: req.schema });
  		walletService
  			.readByKey(userid)
  			.then((results) => {
 ```
 What am I missing?

**Reply by glowkeeper** [reply]
Oooo - `.readByQuery({ sort: 'userid', fields: ['*'] })` - that works! I must be using `.readByKey` incorrectly, right?

**Comment by RageBrahimi** [comment]
```javascript
const Joi = require("joi");

const bodySchema = Joi.object({
  id: Joi.string().required(),
  key: Joi.string().required(),
});

module.exports = function registerEndpoint(router, { services, exceptions }) {
  const { ItemsService } = services;
  const { InvalidPayloadException, ServiceUnavailableException } = exceptions;

  router.post("/key", (req, res, next) => {
    const { error } = bodySchema.validate(req.body);
    if (error) throw new InvalidPayloadException(error.message);

    const { id, key } = req.body;

    const walletService = new ItemsService("wallet", {
      schema: req.schema,
      accountability: req.accountability,
    });

    walletService
      .create({ userid: id, publickey: key })
      .then((results) => res.json(results))
      .catch((error) => {
        return next(new ServiceUnavailableException(error.message));
      });
  });
};
```

[PR request](https://github.com/directus/directus/pull/4098) was approved 2 days ago. Hope this helps! 👍

**Reply by RageBrahimi** [reply]
In terms of the below:

```json
{
  "errors": [
    {
      "message": "You don't have permission to access this.",
      "extensions": {
        "code": "FORBIDDEN"
      }
    }
  ]
}
```

The possible explanation is that your `{"id":"813a4267-a9e7-4949-8044-18ded8782c95" }` does not actually match a user in the `directus_users` table as I can see that field is not populated from your screenshot.

**Reply by glowkeeper** [reply]
Yup - that was indeed the case - I grabbed an id from the wrong server - doh!

**Reply by RageBrahimi** [reply]
Ha! No worries. I would recommend using the above though to validate payload and throw exceptions correctly. Have a good rest of the day!

**Reply by glowkeeper** [reply]
I was _already_ doing exactly that (as I found examples in the directus source)! e.g.
```
router.post('/checkId', (req, res, next) => {

    const userid = req.body.userid

    if ( userid ) {

      const walletService = new ItemsService('wallet', { schema: req.schema });
  	walletService
  	 .readByQuery({ sort: 'userid', fields: ['*'] })
         then((results) => {
          //console.log("got users: ", results)

          let pubkeys = [];
          for (let i = 0; i < results.length; i++) {

            if (results[i].userid == userid ) {
              pubkeys.push(results[i].publickey);
            }
          }

          const userKeys = {
            publickeys: pubkeys
          }
          return res.send(JSON.stringify(userKeys));

        })
        .catch((error) => {

          console.error(error.message);
          return next(new ServiceUnavailableException(error.message));

  	});
    } else {

      console.error("GET key error");
      return next(new ServiceUnavailableException("GET key error"));
    }
});
```
Unfortunately, though, that's not the error I'm talking about at the end of the thread above - I reckon there's something strange going on with `.readByKey` (or I'm using it wrong). I've currently bypassed the error via `.readByQuery` (as you can see above), but that means my code is super inefficient - would be good to get `.readByKey` working!

**Reply by glowkeeper** [reply]
Oooo - no I wasn't (already doing that)! I looked a bit closer - _InvalidPayloadException_ - nice! Thanks, @RageBrahimi !

**Comment by ph-arm** [comment]
I have many route on my custom endpoint, how can I reuse services amongst multiple routes, since empty schema seem to crash? 
(trying to keep code compact and clean)
tldr: noob trying to tidy his code and question might be subpar.
nb: 'module.exports'  crash on my side, have to use 'export default'

**Reply by rijkvanzanten** [reply]
@ph-arm What do you have so far?

The currently `schema` is passed in under `req.schema`, so you can use that in the services setup without having to re-load the schema:

```js
router.get('/my-endpoint', (req, res) => {
  const articles = new ItemsService('articles', { schema: req.schema });
});
```

**Reply by ph-arm** [reply]
as a code squeleton I went with : 

```
export default (router, context) => {
  router.get("/hello-custom-ext", (req, res) => res.send("Hello, World!"));
  router.post("/generate", function (req, res) {}
  router.post("/token", function (req, res) {
      const service = new ItemsService("Communication", {
      schema: req.schema,
      accountability: {
        admin: true,
        role: null,
        permissions: [],
      },
    });
}
}
```

but I would rather share the service across all my route. I hope it is a little more clear. Something like that : 
```
export default (router, context) => {
const service = new ItemsService("Communication", {
      schema: ???req.schema,
      accountability: {
        admin: true,
        role: null,
        permissions: [],
      },
    });
  router.get("/hello-custom-ext", (req, res) => res.send("Hello, World!"));
  router.post("/generate", function (req, res) {}
  router.post("/token", function (req, res) {

}
}
```
I am trying to build a  custom endpoint handling voice/text/mess  with sendgrid/twilio etc..

**Reply by ph-arm** [reply]
Also am I not sure why it is expected for readOne to return forbidden in case of not found. 
Again sorry if it looks like total noob questions.
Already extremely grateful for your work

**Reply by rijkvanzanten** [reply]
> Also am I not sure why it is expected for readOne to return forbidden in case of not found.

It's for security reasons. By returning a 404 instead of a 403, you could extrapolate what items exist in a system, which can leak private data 👍🏻 

That being said, we could/should improve that by having the services return a proper "NOT_FOUND" exception, and only retuning the 403s from the IO layer (eg rest/gql)

**Reply by rijkvanzanten** [reply]
> but I would rather share the service across all my route. I hope it is a little more clear. Something like that :

The schema and accountability will be different for each request, so you will have to create the service every request. What's the reason you want to have the same service outside the route handlers?

**Reply by ph-arm** [reply]
I wrote an endpoint with  many routes, I mostly use admin, because it was a lot simpler and has pretty simple data, fed back by twilio.
In the end, I  put a wrapper around ItemsService.
